File: examples/python/simple_archive.py
"""
Simple archive example - Archive a single URL with HTBase.

Usage:
    python simple_archive.py

Requirements:
    pip install requests
"""

import requests
from typing import Dict, Any


def archive_url(
    url: str,
    item_id: str,
    archiver: str = "readability",
    base_url: str = "http://localhost:8000"
) -> Dict[str, Any]:
    """
    Archive a single URL using HTBase.

    Args:
        url: The URL to archive
        item_id: Unique identifier for this archive
        archiver: Archiver to use (readability, monolith, singlefile-cli, pdf, screenshot, all)
        base_url: HTBase API base URL

    Returns:
        Dictionary with archive result

    Raises:
        requests.exceptions.RequestException: If the request fails
    """
    endpoint = f"{base_url}/api/save/{archiver}"

    payload = {
        "url": url,
        "id": item_id
    }

    response = requests.post(
        endpoint,
        json=payload,
        timeout=300  # 5 minutes
    )

    # Raise exception for HTTP errors
    response.raise_for_status()

    return response.json()


def main():
    """Example usage."""
    # Configuration
    URL_TO_ARCHIVE = "https://example.com"
    ITEM_ID = "example-article-2026-01-09"
    ARCHIVER = "readability"

    print(f"Archiving: {URL_TO_ARCHIVE}")
    print(f"Item ID: {ITEM_ID}")
    print(f"Archiver: {ARCHIVER}")
    print("-" * 50)

    try:
        # Archive the URL
        result = archive_url(
            url=URL_TO_ARCHIVE,
            item_id=ITEM_ID,
            archiver=ARCHIVER
        )

        # Check result
        if result.get("ok") and result.get("exit_code") == 0:
            print("✅ Archive successful!")
            print(f"Saved to: {result['saved_path']}")
            print(f"Database row ID: {result['db_rowid']}")
        else:
            print("❌ Archive failed!")
            print(f"Exit code: {result.get('exit_code')}")
            print(f"Details: {result}")

    except requests.exceptions.Timeout:
        print("❌ Request timed out (archive took >5 minutes)")

    except requests.exceptions.HTTPError as e:
        print(f"❌ HTTP error: {e}")
        print(f"Response: {e.response.text}")

    except requests.exceptions.RequestException as e:
        print(f"❌ Request failed: {e}")


if __name__ == "__main__":
    main()

File: examples/javascript/simple_archive.js
/**
 * Simple archive example - Archive a single URL with HTBase.
 *
 * Usage:
 *   node simple_archive.js
 *
 * Requirements:
 *   npm install node-fetch
 */

const fetch = require('node-fetch');

/**
 * Archive a single URL using HTBase.
 *
 * @param {string} url - The URL to archive
 * @param {string} itemId - Unique identifier for this archive
 * @param {string} archiver - Archiver to use (default: readability)
 * @param {string} baseUrl - HTBase API base URL
 * @returns {Promise<object>} Archive result
 */
async function archiveUrl(
  url,
  itemId,
  archiver = 'readability',
  baseUrl = 'http://localhost:8000'
) {
  const endpoint = `${baseUrl}/api/save/${archiver}`;

  const payload = {
    url: url,
    id: itemId
  };

  const response = await fetch(endpoint, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify(payload),
    timeout: 300000 // 5 minutes
  });

  if (!response.ok) {
    const text = await response.text();
    throw new Error(`HTTP ${response.status}: ${text}`);
  }

  return response.json();
}

/**
 * Main function - example usage.
 */
async function main() {
  // Configuration
  const URL_TO_ARCHIVE = 'https://example.com';
  const ITEM_ID = 'example-article-2026-01-09';
  const ARCHIVER = 'readability';

  console.log(`Archiving: ${URL_TO_ARCHIVE}`);
  console.log(`Item ID: ${ITEM_ID}`);
  console.log(`Archiver: ${ARCHIVER}`);
  console.log('-'.repeat(50));

  try {
    // Archive the URL
    const result = await archiveUrl(URL_TO_ARCHIVE, ITEM_ID, ARCHIVER);

    // Check result
    if (result.ok && result.exit_code === 0) {
      console.log('✅ Archive successful!');
      console.log(`Saved to: ${result.saved_path}`);
      console.log(`Database row ID: ${result.db_rowid}`);
    } else {
      console.log('❌ Archive failed!');
      console.log(`Exit code: ${result.exit_code}`);
      console.log('Details:', result);
    }
  } catch (error) {
    console.error('❌ Error:', error.message);
  }
}

// Run if executed directly
if (require.main === module) {
  main().catch(console.error);
}

// Export for use as module
module.exports = { archiveUrl };
